File: base/android/java/src/org/chromium/base/ScreenStateReceiver.java
// Copyright 2025 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.base;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;

import androidx.annotation.VisibleForTesting;

import org.chromium.base.task.PostTask;
import org.chromium.base.task.TaskRunner;
import org.chromium.base.task.TaskTraits;
import org.chromium.build.annotations.NullMarked;
import org.chromium.build.annotations.Nullable;

/**
 * A {@link BroadcastReceiver} that listens for screen state changes (off/on).
 *
 * <p>This class is a singleton that is never destroyed as it is an application level broadcast
 * receiver. Care should be taken to avoid memory leaks by removing listeners when they are no
 * longer needed.
 */
@NullMarked
public class ScreenStateReceiver extends BroadcastReceiver {
    /** A listener for screen state broadcasts. */
    public interface ScreenStateObserver {
        default void onScreenOff(Context context, Intent intent) {}

        default void onScreenOn(Context context, Intent intent) {}
    }

    // Initialize lazily to facilitate testing.
    private static @Nullable ScreenStateReceiver sInstance;

    private final ObserverList<ScreenStateObserver> mObservers = new ObserverList<>();
    private final TaskRunner mTaskRunner =
            PostTask.createSequencedTaskRunner(TaskTraits.USER_VISIBLE);

    private ScreenStateReceiver() {}

    /**
     * Adds an observer for screen state broadcasts.
     *
     * @param observer The observer to add.
     */
    public static void addObserver(ScreenStateObserver observer) {
        ThreadUtils.assertOnUiThread();
        getInstance().mObservers.addObserver(observer);
    }

    /**
     * Removes an observer for screen state broadcasts.
     *
     * @param observer The observer to remove.
     */
    public static void removeObserver(ScreenStateObserver observer) {
        ThreadUtils.assertOnUiThread();
        getInstance().mObservers.removeObserver(observer);
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        ThreadUtils.assertOnUiThread();
        if (Intent.ACTION_SCREEN_OFF.equals(intent.getAction())) {
            for (ScreenStateObserver observer : mObservers) {
                observer.onScreenOff(context, intent);
            }
        } else if (Intent.ACTION_SCREEN_ON.equals(intent.getAction())) {
            for (ScreenStateObserver observer : mObservers) {
                observer.onScreenOn(context, intent);
            }
        }
    }

    /** Returns the singleton instance of the {@link ScreenStateReceiver}. */
    @VisibleForTesting
    public static ScreenStateReceiver getInstance() {
        ThreadUtils.assertOnUiThread();
        if (sInstance == null) {
            sInstance = new ScreenStateReceiver();
            sInstance.register();
        }
        return sInstance;
    }

    private void register() {
        // Registering a broadcast receiver can be slow. To prevent blocking do this asynchronously.
        mTaskRunner.postDelayedTask(
                () -> {
                    IntentFilter filter = new IntentFilter();
                    filter.addAction(Intent.ACTION_SCREEN_OFF);
                    filter.addAction(Intent.ACTION_SCREEN_ON);
                    ContextUtils.registerProtectedBroadcastReceiver(
                            ContextUtils.getApplicationContext(), ScreenStateReceiver.this, filter);
                },
                0);
    }

    private void unregister() {
        // Unregister the receiver on the same task runner to ensure ordering.
        mTaskRunner.postDelayedTask(
                () -> {
                    ContextUtils.getApplicationContext()
                            .unregisterReceiver(ScreenStateReceiver.this);
                },
                0);
    }

    /** Resets the singleton instance of the {@link ScreenStateReceiver}. */
    public static void resetForTesting() {
        if (sInstance == null) return;

        sInstance.unregister();
        sInstance = null;
    }
}

File: base/android/screen_state_receiver.cc
// Copyright 2026 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "base/android/screen_state_receiver.h"

#include "base/android/jni_android.h"
#include "base/location.h"
#include "base/no_destructor.h"
#include "base/observer_list_threadsafe.h"
#include "third_party/jni_zero/jni_zero.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "base/screen_state_receiver_jni/ScreenStateReceiver_jni.h"

namespace base::android {

namespace {

class ScreenStateReceiverImpl {
 public:
  static ScreenStateReceiverImpl* GetInstance() {
    static base::NoDestructor<ScreenStateReceiverImpl> instance;
    return instance.get();
  }

  ScreenStateReceiverImpl() {
    JNIEnv* env = jni_zero::AttachCurrentThread();
    Java_ScreenStateReceiver_registerThreadSafeNativeScreenStateObserver(env);
  }

  void AddObserver(ScreenStateReceiver::Observer* observer) {
    observers_->AddObserver(observer);
  }

  void RemoveObserver(ScreenStateReceiver::Observer* observer) {
    observers_->RemoveObserver(observer);
  }

  void OnScreenOff() {
    observers_->Notify(FROM_HERE, &ScreenStateReceiver::Observer::OnScreenOff);
  }

  void OnScreenOn() {
    observers_->Notify(FROM_HERE, &ScreenStateReceiver::Observer::OnScreenOn);
  }

 private:
  scoped_refptr<base::ObserverListThreadSafe<ScreenStateReceiver::Observer>>
      observers_ = base::MakeRefCounted<
          base::ObserverListThreadSafe<ScreenStateReceiver::Observer>>();
};

}  // namespace

// static
void ScreenStateReceiver::AddObserver(Observer* observer) {
  ScreenStateReceiverImpl::GetInstance()->AddObserver(observer);
}

// static
void ScreenStateReceiver::RemoveObserver(Observer* observer) {
  ScreenStateReceiverImpl::GetInstance()->RemoveObserver(observer);
}

static void JNI_ScreenStateReceiver_OnScreenOff(JNIEnv* env) {
  ScreenStateReceiverImpl::GetInstance()->OnScreenOff();
}

static void JNI_ScreenStateReceiver_OnScreenOn(JNIEnv* env) {
  ScreenStateReceiverImpl::GetInstance()->OnScreenOn();
}

DEFINE_JNI(ScreenStateReceiver)

}  // namespace base::android
